### PR TITLE
[UNR-2796] Enhance deployment startup report

### PR DESCRIPTION
**Pull request by ImprobableNic**
#### Description
Given a configuration where some workers would need to be manually started, the Cloud Deployment launcher reports that the deployment won't start any worker.
The message reported has been changed to tell which workers won't be started by parsing the configuration file.

#### Tests
Launching Cloud Deploy in the example project, with the one_worker.json config file should report that the DeploymentManager worker needs to be manually started.

#### Documentation

### CODE REVIEW

File: SpatialGDK/Source/SpatialGDKEditor/Private/SpatialGDKEditorSettings.cpp [hunk 1]
```diff
@@ -226,21 +229,74 @@ void USpatialGDKEditorSettings::SetNumberOfSimulatedPlayers(uint32 Number)
...
+	{
+		FArchive* ConfigFile = IFileManager::Get().CreateFileReader(*LaunchConfigPath);
+
+		if (!ConfigFile)
+		{
+			UE_LOG(LogSpatialEditorSettings, Error, TEXT("Could not open configuration file %s"), *LaunchConfigPath);
+			return false;
+		}
+
+		auto JsonReader = TJsonReader<char>::Create(ConfigFile);
```
Comment (joshuahuburn) [inline SpatialGDK/Source/SpatialGDKEditor/Private/SpatialGDKEditorSettings.cpp]: I think we have a style rule against using `auto` unless it's in a loop. Can we change this to the full type please.

File: SpatialGDK/Source/SpatialGDKEditor/Private/SpatialGDKEditorSettings.cpp [hunk 2]
```diff
@@ -226,21 +229,74 @@ void USpatialGDKEditorSettings::SetNumberOfSimulatedPlayers(uint32 Number)
 	SaveConfig();
 }
 
-bool USpatialGDKEditorSettings::IsManualWorkerConnectionSet(const FString& LaunchConfigPath)
+bool USpatialGDKEditorSettings::IsManualWorkerConnectionSet(const FString& LaunchConfigPath, TArray<FString>& OutWorkersManuallyLaunched)
 {
-	FString FileContents;
-	FFileHelper::LoadFileToString(FileContents, *LaunchConfigPath);
+	TSharedPtr<FJsonValue> Document;
```
Comment (joshuahuburn) [inline SpatialGDK/Source/SpatialGDKEditor/Private/SpatialGDKEditorSettings.cpp]:
```suggestion
	TSharedPtr<FJsonValue> LaunchConfigJson;
```

File: SpatialGDK/Source/SpatialGDKEditor/Private/SpatialGDKEditorSettings.cpp [hunk 3]
```diff
@@ -226,21 +229,74 @@ void USpatialGDKEditorSettings::SetNumberOfSimulatedPlayers(uint32 Number)
...
+		}
+
+		auto JsonReader = TJsonReader<char>::Create(ConfigFile);
+
+		FJsonSerializer::Deserialize(*JsonReader, Document);
+
+		delete ConfigFile;
+		ConfigFile = nullptr;
+	}
+	const TSharedPtr<FJsonObject>* DocRoot;
```
Comment (joshuahuburn) [inline SpatialGDK/Source/SpatialGDKEditor/Private/SpatialGDKEditorSettings.cpp]:
```suggestion
	const TSharedPtr<FJsonObject>* LaunchConfigJsonRootObject;
```
Or something

File: SpatialGDK/Source/SpatialGDKEditor/Private/SpatialGDKEditorSettings.cpp [hunk 4]
```diff
@@ -226,21 +229,74 @@ void USpatialGDKEditorSettings::SetNumberOfSimulatedPlayers(uint32 Number)
...
+	if (!Document || !Document->TryGetObject(DocRoot))
+	{
+		UE_LOG(LogSpatialEditorSettings, Error, TEXT("Invalid configuration file %s"), *LaunchConfigPath);
+		return false;
+	}
 
-	const FRegexPattern ManualWorkerFlagPattern("\"manual_worker_connection_only\" *: *true");
-	FRegexMatcher ManualWorkerFlagMatcher(ManualWorkerFlagPattern, FileContents);
+	const TSharedPtr<FJsonObject>* LoadBalancingField;
+		if (!(*DocRoot)->TryGetObjectField("load_balancing", LoadBalancingField))
```
Comment (joshuahuburn) [inline SpatialGDK/Source/SpatialGDKEditor/Private/SpatialGDKEditorSettings.cpp]: Spacing is off here.

File: SpatialGDK/Source/SpatialGDKEditor/Private/SpatialGDKEditorSettings.cpp [hunk 5]
```diff
@@ -226,21 +229,74 @@ void USpatialGDKEditorSettings::SetNumberOfSimulatedPlayers(uint32 Number)
...
+			return false;
+		}
+
+		auto JsonReader = TJsonReader<char>::Create(ConfigFile);
+
+		FJsonSerializer::Deserialize(*JsonReader, Document);
+
+		delete ConfigFile;
+		ConfigFile = nullptr;
+	}
```
Comment (joshuahuburn) [inline SpatialGDK/Source/SpatialGDKEditor/Private/SpatialGDKEditorSettings.cpp]:
```suggestion
	}
	
```
I think we usually have a newline after `}`

File: SpatialGDK/Source/SpatialGDKEditor/Private/SpatialGDKEditorSettings.cpp [hunk 6]
```diff
@@ -226,21 +229,75 @@ void USpatialGDKEditorSettings::SetNumberOfSimulatedPlayers(uint32 Number)
...
+	if (!LaunchConfigJson || !LaunchConfigJson->TryGetObject(LaunchConfigJsonRootObject))
+	{
+		UE_LOG(LogSpatialEditorSettings, Error, TEXT("Invalid configuration file %s"), *LaunchConfigPath);
+		return false;
+	}
+
+	const TSharedPtr<FJsonObject>* LoadBalancingField;
+	if (!(*LaunchConfigJsonRootObject)->TryGetObjectField("load_balancing", LoadBalancingField))
+	{
+		return false;
```
Comment (jessicafalk) [inline SpatialGDK/Source/SpatialGDKEditor/Private/SpatialGDKEditorSettings.cpp]: can we log some error here as well to indicate where the problem is?

File: SpatialGDK/Source/SpatialGDKEditor/Private/SpatialGDKEditorSettings.cpp [hunk 7]
```diff
@@ -226,21 +229,75 @@ void USpatialGDKEditorSettings::SetNumberOfSimulatedPlayers(uint32 Number)
...
-	const FRegexPattern ManualWorkerFlagPattern("\"manual_worker_connection_only\" *: *true");
-	FRegexMatcher ManualWorkerFlagMatcher(ManualWorkerFlagPattern, FileContents);
+	const TArray<TSharedPtr<FJsonValue>>* LayerConfigurations;
 
-	if (ManualWorkerFlagMatcher.FindNext())
+	if (!(*LoadBalancingField)->TryGetArrayField("layer_configurations", LayerConfigurations))
 	{
-		UE_LOG(LogSpatialEditorSettings, Warning, TEXT("Launch configuration for cloud deployment has \"manual_worker_connection_only\" set to true. This means server workers will need to be connected manually."));
-		return true;
+		return false;
```
Comment (jessicafalk) [inline SpatialGDK/Source/SpatialGDKEditor/Private/SpatialGDKEditorSettings.cpp]: same here

File: SpatialGDK/Source/SpatialGDKEditor/Private/SpatialGDKEditorSettings.cpp (hunk 6 above)
Comment (ImprobableNic) [reply SpatialGDK/Source/SpatialGDKEditor/Private/SpatialGDKEditorSettings.cpp]: Is it an error to not have a load_balancing section ?

File: SpatialGDK/Source/SpatialGDKEditor/Private/SpatialGDKEditorSettings.cpp [hunk 8]
```diff
@@ -226,21 +229,74 @@ void USpatialGDKEditorSettings::SetNumberOfSimulatedPlayers(uint32 Number)
 	SaveConfig();
 }
 
-bool USpatialGDKEditorSettings::IsManualWorkerConnectionSet(const FString& LaunchConfigPath)
+bool USpatialGDKEditorSettings::IsManualWorkerConnectionSet(const FString& LaunchConfigPath, TArray<FString>& OutWorkersManuallyLaunched)
 {
-	FString FileContents;
-	FFileHelper::LoadFileToString(FileContents, *LaunchConfigPath);
+	TSharedPtr<FJsonValue> Document;
+	{
```
[the remark this answers is not in the document]
Comment (ImprobableNic) [reply SpatialGDK/Source/SpatialGDKEditor/Private/SpatialGDKEditorSettings.cpp]: We do not, but I prefer to scope variables that won't be used for the rest of the function, in that case the FArchive pointer.

File: SpatialGDK/Source/SpatialGDKEditor/Private/SpatialGDKEditorSettings.cpp [hunk 9]
```diff
@@ -226,21 +229,74 @@ void USpatialGDKEditorSettings::SetNumberOfSimulatedPlayers(uint32 Number)
...
 	}
 
-	return false;
+	for (const auto& LayerConfigurationValue : *LayerConfigurations)
+	{
+		if (const TSharedPtr<FJsonObject> LayerConfiguration = LayerConfigurationValue->AsObject())
+		{
+			const TSharedPtr<FJsonObject>* OptionsField;
+			if (LayerConfiguration->TryGetObjectField("options", OptionsField))
+			{
```
[the remark this answers is not in the document]
Comment (ImprobableNic) [reply SpatialGDK/Source/SpatialGDKEditor/Private/SpatialGDKEditorSettings.cpp]: The succession of 
`if(!...)
{
  continue;
}`
made my eyes bleed at the time.
I grouped several checks in a single conditional to reduce nesting, what do you think ?

File: SpatialGDK/Source/SpatialGDKEditor/Private/SpatialGDKEditorSettings.cpp [hunk 10]
```diff
@@ -226,21 +229,72 @@ void USpatialGDKEditorSettings::SetNumberOfSimulatedPlayers(uint32 Number)
...
 }
 
-bool USpatialGDKEditorSettings::IsManualWorkerConnectionSet(const FString& LaunchConfigPath)
+bool USpatialGDKEditorSettings::IsManualWorkerConnectionSet(const FString& LaunchConfigPath, TArray<FString>& OutWorkersManuallyLaunched)
 {
-	FString FileContents;
-	FFileHelper::LoadFileToString(FileContents, *LaunchConfigPath);
+	TSharedPtr<FJsonValue> LaunchConfigJson;
+	{
+		FArchive* ConfigFile = IFileManager::Get().CreateFileReader(*LaunchConfigPath);
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDKEditor/Private/SpatialGDKEditorSettings.cpp]: Wrap in a TUniquePtr

File: SpatialGDK/Source/SpatialGDKEditor/Private/SpatialGDKEditorSettings.cpp (hunk 6 above)
Comment (joshuahuburn) [reply SpatialGDK/Source/SpatialGDKEditor/Private/SpatialGDKEditorSettings.cpp]: No it's not an error

**Pull request merged**